### ISSUE TITLE
Using `open` with some text file trims the last "\n".

### ISSUE THREAD

**Issue by maxim-uvarov** [nushell/nushell#13155]
### Describe the bug

Using `open` with some text file trims the last "\n".
I opened the `test.txt` created with `"\n\n" | save test.txt`
And it has two `new line` characters in it.
`open -r` produces the same result.

### How to reproduce

```nu
> "\n\n" | str length
2

> "\n\n" | save test.txt

> open test.txt



> open test.txt | str length
1

> open test.txt -r | str length
1

> open test.txt | str distance "\n\n"
1
```

Yet this works as expected:
```
> open test.txt -r | into binary
Length: unknown (stream) | printable whitespace ascii_other non_ascii
00000000:   0a 0a                                                __

> open test.txt | save test2.txt
> code test2.txt # it has two new lines in it in the editor

```

### Expected behavior

I expect nu not to alter the content of a file.

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.94.2                                          |
| major              | 0                                               |
| minor              | 94                                              |
| patch              | 2                                               |
| branch             | main                                            |
| commit_hash        | be8c1dc0066cd1034a6b110a622f47b516bfe029        |
| build_os           | macos-aarch64                                   |
| build_target       | aarch64-apple-darwin                            |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)             |
| rust_channel       | 1.77.2-aarch64-apple-darwin                     |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)             |
| build_time         | 2024-06-04 08:20:37 +00:00                      |
| build_rust_channel | release                                         |
| allocator          | mimalloc                                        |
| features           | default, sqlite, system-clipboard, trash, which |
| installed_plugins  | explore, image, polars, regex                   |

**Comment by userwiths** [comment]
I believe this an issue with the `str` functionality and not with `open`

I'm running the `main` branch locally and get this experience
```bash
vim test.txt # enter 2 new lines (\n) and exit
"\n\n" | save test2.txt # described case

cat test.txt | wc -c # returns 2
cat test2.txt | wc -c # returns 2
open test.txt | wc -c # returns 2
open test2.txt | wc -c # returns 2

open test.txt | str length # returns 1
open test2.txt | str lenght # returns 1
cat test.txt | str length # returns 1
cat test2.txt | str lenght # returns 1
```

Regardless if I use `cat` or `open` , `wc` reports 2 characters. And, again, regardless of the printing used `str length` reports a single character.

Might be some time till I can look at the code, but thats what the testing found.

**Comment by userwiths** [comment]
Ok, seems the pipe is the issue as `str length` receives a single newline.

So this is the function that does the trimming:
https://github.com/nushell/nushell/blob/6e1e824473e15eba246e1c43704c5d88fa237a17/crates/nu-protocol/src/pipeline/byte_stream.rs#L956

And its supposed to **not** trim the ending `\n` if the stream's source is `Child` or `File`, but the `is_external` function seems to return `true` for the examples above. I'm having problems understanding where exactly we create and assign the type of stream tho :thinking: 

Specific example enters the `into_value` function and get trimmed there:
https://github.com/nushell/nushell/blob/6e1e824473e15eba246e1c43704c5d88fa237a17/crates/nu-protocol/src/pipeline/byte_stream.rs#L529

Just to avoid suspicion, it behaves the same way on a system with `open` (from the Tcl Build-In commands) and on a system without `open` (using the `nushell` version of the command).

I'm leaving this to someone that can make sense of the streams, best of luck.

**Comment by maxim-uvarov** [comment]
thank you, @userwiths for investigating!

I found that after using `collect` an ending new line disapears (which isn't desired):

```
> (char lf) + 'a' + (char lf) | save test.txt -f

> open test.txt | into binary
Length: unknown (stream) | printable whitespace ascii_other non_ascii
00000000:   0a 61 0a                                             _a_

> open test.txt | collect | into binary
Length: 2 (0x2) bytes | printable whitespace ascii_other non_ascii
00000000:   0a 61
```

<img width="729" alt="image" src="https://github.com/user-attachments/assets/9a2ce12c-331d-4012-a6bf-561aaec0f2f5">

**Comment by isti115** [comment]
Assuming that this will be fixed later on (I might even look into it myself if I'll have the time), I didn't want to hardcode adding the extra '\n' into my script, so I tried to look for another way to get a `string` out of a `stream` without losing the trailing newline, and found that `open test.txt --raw | decode` is (at least in my case) a viable workaround.

![image](https://github.com/user-attachments/assets/1ff77b50-7348-40e5-b0b8-19842c684e94)

**Comment by fdncred** [comment]
If we change the `is_external()` function in byte_stream.rs to this it seems to work for this scenario. (removed the ByteStreamSource::File() in the matches!)
```nushell
    /// Source is a `Child` or `File`, rather than `Read`. Currently affects trimming
    fn is_external(&self) -> bool {
        matches!(self, ByteStreamSource::Child(..))
    }
```
![image](https://github.com/user-attachments/assets/5d398ff9-8b19-48b1-9eb6-2a8abb204f35)

**Comment by IanManske** [comment]
I believe the original reasoning for trimming trailing new lines was:
1. Users coming from POSIX shells are used to command substitution stripping trailing new lines.
2. Trimming ending newlines from external commands is a common occurrence, so it might be nice to have it as the default behavior when collecting into, e.g., a variable.

But if this is causing more confusion that it's worth, then perhaps we should just remove it. Notably, we only strip the last newline, but POSIX strips ALL trailing newlines. Also, `lines` already strips the last newline (see [`str::lines`](https://doc.rust-lang.org/std/primitive.str.html#method.lines)), so perhaps this or `str trim` should be used instead. 

Either way, commands that collect their input should not be stripping newlines (e.g., `str length` above). The stripping was only supposed to happen if the external command or `open` was the last pipeline command. E.g., `let x = some_cmd` or `$'(some_cmd)'`

In the mean time, the current workaround is to put `into string` or `into binary` after `open` or an external command.

```
> open test.txt | into binary
Length: unknown (stream) | printable whitespace ascii_other non_ascii
00000000:   61 0a 62 0a                                          a_b_
```

**Comment by maxim-uvarov** [comment]
I'm not sure, but maybe this inconsistency of `to text` and `str join (char nl)` is related:

```
> [a] | each {} | str join (char nl) | into binary                                                       
Length: unknown (stream) | printable whitespace ascii_other non_ascii
00000000:   61

> [a] | each {} | to text | into binary                      
Length: unknown (stream) | printable whitespace ascii_other non_ascii
00000000:   61 0a
```

**Comment by weirdan** [comment]
@maxim-uvarov probably not, `to text` just explicitly adds a new line after every item:
https://github.com/nushell/nushell/blob/e911ff4d67e5342f325d825745f6f5c01504d6e6/crates/nu-command/src/formats/to/text.rs#L56

while `str join` adds separators _between_ items.

**Comment by maxim-uvarov** [comment]
I see, just because it handles `stream` and `string` differently
https://github.com/nushell/nushell/blob/e911ff4d67e5342f325d825745f6f5c01504d6e6/crates/nu-command/src/formats/to/text.rs#L45-L58

thank you so much

**Comment by weirdan** [comment]
> but why to text doesn't do that with simple list without `each {}`

Sounds like a (separate) bug to me.

**Comment by fdncred** [comment]
easy enough to fix. should they both be with a line ending or without a line ending? from my experience, consistency is better, but this is kind of a no-win situation because people will complain with either situation.

**Comment by weirdan** [comment]
> should they both be with a line ending or without a line ending?

If you're talking about the `to text` stream / list discrepancy, then I'd vote for having the trailing newline in both cases. `to text` is likely to be used before `| save`, and it's customary (at least in Unix world) to have text files to end with a newline. Also it makes it easier to concatenate results of two `to text` calls. I'd expect the following:
```nushell
([a b] | to text) ++ ([x z] | to text)
```
to result in
```
a
b
x
z
```
while currently it produces
```
a
bx
z
```

**Comment by maxim-uvarov** [comment]
Maybe we make the trailing new line by default and just add the flag `--no-newline` as the `nu` cli has it?

**Comment by fdncred** [comment]
I think this is better https://github.com/nushell/nushell/pull/14158, but maybe not perfect yet.

**Comment by maxim-uvarov** [comment]
I found this issue fixed in 0.100.0
```
>  "\n\n" | save test.txt

> open test.txt | str length
2

> open test.txt -r | str length
2

> open test.txt | str distance "\n\n"
0
```

Thanks everyone!
